MULE-19530: Improve documentation about MetadataScopes and MetadataKeyId (#128)
mulesoft/docs-mule-sdk: modules/ROOT/pages/metadata-keys.adoc

// _TODO Fix broken Image
// image::<missing>metadata/keys_single_drop.png[MetadataKeys Single Level Dropdown]

> After the resolver is defined, you can use it as an operation or source by adding a `@MetadataScope` annotation with its `keysResolver` value at the class definition level, while also annotating the parameter that is the key with `@MetadataKeyId`.

Reviewer: ```suggestion
After the resolver is defined, you can use it as an operation or source by adding a `@MetadataScope` annotation with its `keysResolver` value at the class definition level and also annotating the parameter that is the key with `@MetadataKeyId`.
```

  }
----

[NOTE]

> The example described above is the only way to support metadata keys for sources. Do not set the resolver at the `@MetadataKeyId` annotation when defining sources.

Reviewer: ```suggestion
The previous example is the only way to support metadata keys for sources. Do not set the resolver at the `@MetadataKeyId` annotation when defining sources.
```

> Setting the `@MetadataScope` annotation at the operation class definition level indicates that all of its operations provide metadata and that they must all have a parameter annotated with `@MetadataKeyId` while they inherit the resolver from the scope.

Reviewer: ```suggestion
Setting the `@MetadataScope` annotation at the operation class definition level indicates that all of its operations provide metadata and that they must all have a parameter annotated with `@MetadataKeyId` because they inherit the resolver from the scope.
```

Reviewer: ?

Reviewer 2: Dev confirmed this makes sense and does not change the meaning.

> However, you can also set a key resolver for a single operation:

Reviewer: ```suggestion
You can also set a key resolver for a single operation:
```

[source,java,linenums]
----
public class MyOperations {

    public void operationWithMetadataKey(@MetadataKeyId(RecordKeysResolver.class) String type,
                       @Config MyConfig config){
        //...
    }

    public void plainOperation(@Config MyConfig config){
        //...
    }
}
----

In addition, if there is a `@MetadataScope` annotation at the operation class definition level, you can set a value for the `@MetadataKeyId` annotation in a parameter to override the resolver defined at the scope level for a single operation:

[source,java,linenums]
----
@MetadataScope(keysResolver = ClassLevelResolver.class)
public class MyOperations {

    public void operation(@MetadataKeyId String type, @Config MyConfig config){
        // This operation will use the ClassLevelResolver
    }

    public void operationWithAnotherResolver(@MetadataKeyId(OverridingResolver.class) String type,
                       @Config MyConfig config){
        // This operation will use the OverridingResolver
    }
}
----

[IMPORTANT]

> The two examples described above (setting the resolver at the parameter level and overriding the inherited resolver from the scope) work only when defining operations. These features are not available for sources.

Reviewer: ```suggestion
The two previous examples (setting the resolver at the parameter level and overriding the inherited resolver from the scope) work only when defining operations. These features are not available for sources.
```
